File: site/src/contracts/failure-discovery.ts
export const FAILURE_DISCOVERY_KINDS = ["structured_failure", "case_exclusion"] as const;
export const FAILURE_DISCOVERY_DISPOSITIONS = [
  "excluded", "warning", "conditional", "observed_failure", "unsupported",
] as const;

export type FailureDiscoveryKind = typeof FAILURE_DISCOVERY_KINDS[number];
export type FailureDiscoveryDisposition = typeof FAILURE_DISCOVERY_DISPOSITIONS[number];

export interface FailureDiscoveryMitigation {
  action: string;
  applicability: string;
  tradeoff: string;
}

export interface FailureCaseContext {
  question: string;
  decision_variables: string;
  objective: string;
  constraints: string;
}

export interface FailureDiscoveryEntry {
  entry_id: string;
  entry_kind: FailureDiscoveryKind;
  disposition: FailureDiscoveryDisposition;
  title_ja: string;
  title_en: string;
  summary: string;
  scope: "method_theory" | "implementation_specific" | "mixed" | "case_specific";
  severity: "critical" | "high" | "warning" | "info" | "not_applicable";
  recoverability: "recoverable" | "conditional" | "fatal" | "not_applicable";
  confidence: "high" | "medium" | "low" | "unverified" | "not_applicable";
  failure_mode_id: string | null;
  case_id: string | null;
  method_ids: string[];
  implementation_ids: string[];
  feature_ids: string[];
  scenario_ids: string[];
  source_ids: string[];
  symptoms: string[];
  diagnostics: string[];
  mitigations: FailureDiscoveryMitigation[];
  related_failure_mode_ids: string[];
  case_context: FailureCaseContext | null;
  canonical_route: string;
  last_verified: string;
}

export interface FailureDiscoveryIndex {
  contract_version: "1.0.0";
  dataset_version: string;
  generated_at: string;
  summary: {
    total_entries: number;
    structured_failure_count: number;
    case_exclusion_count: number;
    entries_with_scenarios: number;
  };
  entries: FailureDiscoveryEntry[];
}

export function parseFailureDiscoveryIndex(value: unknown): FailureDiscoveryIndex {
  const data = record(value, "failure discovery index");
  exactKeys(
    data,
    ["contract_version", "dataset_version", "generated_at", "summary", "entries"],
    "failure discovery index",
  );
  if (data.contract_version !== "1.0.0") throw new Error("Unsupported failure discovery contract.");
  const entries = array(data.entries, "entries").map(parseEntry);
  const ids = entries.map((entry) => entry.entry_id);
  if (new Set(ids).size !== ids.length || ids.some((id, index) => index > 0 && id < ids[index - 1]!)) {
    throw new Error("Failure discovery entries must have unique sorted IDs.");
  }
  const summaryData = record(data.summary, "summary");
  exactKeys(
    summaryData,
    ["total_entries", "structured_failure_count", "case_exclusion_count", "entries_with_scenarios"],
    "summary",
  );
  const summary = {
    total_entries: integer(summaryData.total_entries, "total_entries"),
    structured_failure_count: integer(summaryData.structured_failure_count, "structured_failure_count"),
    case_exclusion_count: integer(summaryData.case_exclusion_count, "case_exclusion_count"),
    entries_with_scenarios: integer(summaryData.entries_with_scenarios, "entries_with_scenarios"),
  };
  const expected = {
    total_entries: entries.length,
    structured_failure_count: entries.filter((entry) => entry.entry_kind === "structured_failure").length,
    case_exclusion_count: entries.filter((entry) => entry.entry_kind === "case_exclusion").length,
    entries_with_scenarios: entries.filter((entry) => entry.scenario_ids.length > 0).length,
  };
  if (JSON.stringify(summary) !== JSON.stringify(expected)) {
    throw new Error("Failure discovery summary must be derived from entries.");
  }
  const failureIds = new Set(entries.flatMap((entry) => entry.failure_mode_id ? [entry.failure_mode_id] : []));
  for (const entry of entries) {
    const missing = entry.related_failure_mode_ids.filter((id) => !failureIds.has(id));
    if (missing.length) throw new Error(`Failure discovery entry has dangling failure links: ${missing.join(", ")}.`);
  }
  return {
    contract_version: "1.0.0",
    dataset_version: text(data.dataset_version, "dataset_version"),
    generated_at: text(data.generated_at, "generated_at"),
    summary,
    entries,
  };
}

function parseEntry(value: unknown, index: number): FailureDiscoveryEntry {
  const field = `entries[${index}]`;
  const data = record(value, field);
  exactKeys(data, [
    "entry_id", "entry_kind", "disposition", "title_ja", "title_en", "summary", "scope",
    "severity", "recoverability", "confidence", "failure_mode_id", "case_id", "method_ids",
    "implementation_ids", "feature_ids", "scenario_ids", "source_ids", "symptoms", "diagnostics",
    "mitigations", "related_failure_mode_ids", "case_context", "canonical_route", "last_verified",
  ], field);
  const entry_kind = enumValue(data.entry_kind, FAILURE_DISCOVERY_KINDS, `${field}.entry_kind`);
  const disposition = enumValue(data.disposition, FAILURE_DISCOVERY_DISPOSITIONS, `${field}.disposition`);
  const scope = enumValue(
    data.scope,
    ["method_theory", "implementation_specific", "mixed", "case_specific"] as const,
    `${field}.scope`,
  );
  const severity = enumValue(
    data.severity,
    ["critical", "high", "warning", "info", "not_applicable"] as const,
    `${field}.severity`,
  );
  const recoverability = enumValue(
    data.recoverability,
    ["recoverable", "conditional", "fatal", "not_applicable"] as const,
    `${field}.recoverability`,
  );
  const confidence = enumValue(
    data.confidence,
    ["high", "medium", "low", "unverified", "not_applicable"] as const,
    `${field}.confidence`,
  );
  const failure_mode_id = nullableText(data.failure_mode_id, `${field}.failure_mode_id`);
  const case_id = nullableText(data.case_id, `${field}.case_id`);
  const case_context = data.case_context === null ? null : parseContext(data.case_context, `${field}.case_context`);
  const canonical_route = text(data.canonical_route, `${field}.canonical_route`);
  if (!canonical_route.startsWith("/failures?entry=")) throw new Error(`${field}.canonical_route is invalid.`);
  const entry: FailureDiscoveryEntry = {
    entry_id: text(data.entry_id, `${field}.entry_id`), entry_kind, disposition,
    title_ja: text(data.title_ja, `${field}.title_ja`), title_en: text(data.title_en, `${field}.title_en`),
    summary: text(data.summary, `${field}.summary`), scope, severity, recoverability, confidence,
    failure_mode_id, case_id,
    method_ids: uniqueStrings(data.method_ids, `${field}.method_ids`),
    implementation_ids: uniqueStrings(data.implementation_ids, `${field}.implementation_ids`),
    feature_ids: uniqueStrings(data.feature_ids, `${field}.feature_ids`),
    scenario_ids: uniqueStrings(data.scenario_ids, `${field}.scenario_ids`),
    source_ids: uniqueStrings(data.source_ids, `${field}.source_ids`),
    symptoms: uniqueStrings(data.symptoms, `${field}.symptoms`),
    diagnostics: uniqueStrings(data.diagnostics, `${field}.diagnostics`),
    mitigations: array(data.mitigations, `${field}.mitigations`).map((item, mitigationIndex) => {
      const mitigation = record(item, `${field}.mitigations[${mitigationIndex}]`);
      exactKeys(mitigation, ["action", "applicability", "tradeoff"], `${field}.mitigations[${mitigationIndex}]`);
      return {
        action: text(mitigation.action, "action"),
        applicability: text(mitigation.applicability, "applicability"),
        tradeoff: text(mitigation.tradeoff, "tradeoff"),
      };
    }),
    related_failure_mode_ids: uniqueStrings(data.related_failure_mode_ids, `${field}.related_failure_mode_ids`),
    case_context, canonical_route, last_verified: text(data.last_verified, `${field}.last_verified`),
  };
  if (entry_kind === "structured_failure") {
    if (!failure_mode_id || case_id || case_context || scope === "case_specific" || severity === "not_applicable") {
      throw new Error(`${field} has inconsistent structured-failure semantics.`);
    }
    if (!entry.symptoms.length || !entry.diagnostics.length || !entry.mitigations.length) {
      throw new Error(`${field} requires symptoms, diagnostics, and mitigations.`);
    }
  } else if (!case_id || failure_mode_id || !case_context || entry.method_ids.length !== 1
    || disposition !== "excluded" || scope !== "case_specific"
    || severity !== "not_applicable" || recoverability !== "not_applicable"
    || confidence !== "not_applicable") {
    throw new Error(`${field} has inconsistent Case-exclusion semantics.`);
  }
  return entry;
}

function parseContext(value: unknown, field: string): FailureCaseContext {
  const data = record(value, field);
  exactKeys(data, ["question", "decision_variables", "objective", "constraints"], field);
  return {
    question: text(data.question, `${field}.question`),
    decision_variables: text(data.decision_variables, `${field}.decision_variables`),
    objective: text(data.objective, `${field}.objective`),
    constraints: text(data.constraints, `${field}.constraints`),
  };
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) throw new Error(`${field} must be an object.`);
  return value as Record<string, unknown>;
}
function array(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(`${field} must be an array.`);
  return value;
}
function text(value: unknown, field: string): string {
  if (typeof value !== "string" || !value.trim()) throw new Error(`${field} must be non-empty.`);
  return value;
}
function nullableText(value: unknown, field: string): string | null { return value === null ? null : text(value, field); }
function integer(value: unknown, field: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) throw new Error(`${field} must be a non-negative integer.`);
  return value;
}
function uniqueStrings(value: unknown, field: string): string[] {
  const result = array(value, field).map((item, index) => text(item, `${field}[${index}]`));
  if (new Set(result).size !== result.length || result.some((item, index) => index > 0 && item < result[index - 1]!)) {
    throw new Error(`${field} must be unique and sorted.`);
  }
  return result;
}
function enumValue<const T extends readonly string[]>(value: unknown, allowed: T, field: string): T[number] {
  if (typeof value !== "string" || !allowed.includes(value)) throw new Error(`${field} is invalid.`);
  return value as T[number];
}
function exactKeys(data: Record<string, unknown>, expected: readonly string[], field: string): void {
  const allowed = new Set(expected);
  const unknown = Object.keys(data).filter((key) => !allowed.has(key));
  const missing = expected.filter((key) => !Object.prototype.hasOwnProperty.call(data, key));
  if (unknown.length) throw new Error(`${field} has unknown fields: ${unknown.join(", ")}.`);
  if (missing.length) throw new Error(`${field} is missing fields: ${missing.join(", ")}.`);
}

File: src/optimization_compass/failure_discovery.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Self
from urllib.parse import quote

from pydantic import BaseModel, ConfigDict, Field, model_validator

from optimization_compass.db import KnowledgeRepository


class FailureDiscoveryModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class FailureMitigation(FailureDiscoveryModel):
    action: str = Field(min_length=1)
    applicability: str = Field(min_length=1)
    tradeoff: str = Field(min_length=1)


class CaseContext(FailureDiscoveryModel):
    question: str = Field(min_length=1)
    decision_variables: str = Field(min_length=1)
    objective: str = Field(min_length=1)
    constraints: str = Field(min_length=1)


class FailureDiscoveryEntry(FailureDiscoveryModel):
    entry_id: str = Field(min_length=3)
    entry_kind: Literal["structured_failure", "case_exclusion"]
    disposition: Literal["excluded", "warning", "conditional", "observed_failure", "unsupported"]
    title_ja: str = Field(min_length=1)
    title_en: str = Field(min_length=1)
    summary: str = Field(min_length=1)
    scope: Literal["method_theory", "implementation_specific", "mixed", "case_specific"]
    severity: Literal["critical", "high", "warning", "info", "not_applicable"]
    recoverability: Literal["recoverable", "conditional", "fatal", "not_applicable"]
    confidence: Literal["high", "medium", "low", "unverified", "not_applicable"]
    failure_mode_id: str | None
    case_id: str | None
    method_ids: list[str]
    implementation_ids: list[str]
    feature_ids: list[str]
    scenario_ids: list[str]
    source_ids: list[str]
    symptoms: list[str]
    diagnostics: list[str]
    mitigations: list[FailureMitigation]
    related_failure_mode_ids: list[str]
    case_context: CaseContext | None
    canonical_route: str = Field(pattern=r"^/failures\?entry=")
    last_verified: str = Field(min_length=1)

    @model_validator(mode="after")
    def validate_semantics(self) -> Self:
        list_fields = (
            self.method_ids,
            self.implementation_ids,
            self.feature_ids,
            self.scenario_ids,
            self.source_ids,
            self.symptoms,
            self.diagnostics,
            self.related_failure_mode_ids,
        )
        if any(values != sorted(set(values)) for values in list_fields):
            raise ValueError("failure discovery lists must be unique and sorted")
        if self.entry_kind == "structured_failure":
            if (
                self.failure_mode_id is None
                or self.case_id is not None
                or self.case_context is not None
            ):
                raise ValueError("structured failure identity is inconsistent")
            if not self.symptoms or not self.diagnostics or not self.mitigations:
                raise ValueError(
                    "structured failures require symptoms, diagnostics, and mitigations"
                )
            if self.scope == "case_specific" or self.severity == "not_applicable":
                raise ValueError("structured failure scope/severity is inconsistent")
        else:
            if (
                self.case_id is None
                or self.failure_mode_id is not None
                or self.case_context is None
            ):
                raise ValueError("case exclusion identity is inconsistent")
            if len(self.method_ids) != 1 or self.disposition != "excluded":
                raise ValueError("case exclusions require one excluded method")
            if self.scope != "case_specific":
                raise ValueError("case exclusions must use case_specific scope")
            if any(
                value != "not_applicable"
                for value in (self.severity, self.recoverability, self.confidence)
            ):
                raise ValueError(
                    "case exclusions do not infer severity, recoverability, or confidence"
                )
        return self


class FailureDiscoverySummary(FailureDiscoveryModel):
    total_entries: int = Field(ge=0)
    structured_failure_count: int = Field(ge=0)
    case_exclusion_count: int = Field(ge=0)
    entries_with_scenarios: int = Field(ge=0)


class FailureDiscoveryIndex(FailureDiscoveryModel):
    contract_version: Literal["1.0.0"] = "1.0.0"
    dataset_version: str = Field(min_length=1)
    generated_at: datetime
    summary: FailureDiscoverySummary
    entries: list[FailureDiscoveryEntry]

    @model_validator(mode="after")
    def validate_index(self) -> Self:
        ids = [entry.entry_id for entry in self.entries]
        if ids != sorted(ids) or len(ids) != len(set(ids)):
            raise ValueError("failure discovery entries must be unique and sorted")
        structured_ids = {
            entry.failure_mode_id for entry in self.entries if entry.failure_mode_id is not None
        }
        dangling = {
            related
            for entry in self.entries
            for related in entry.related_failure_mode_ids
            if related not in structured_ids
        }
        if dangling:
            raise ValueError(f"failure discovery has dangling failure links: {sorted(dangling)}")
        expected = FailureDiscoverySummary(
            total_entries=len(self.entries),
            structured_failure_count=sum(
                entry.entry_kind == "structured_failure" for entry in self.entries
            ),
            case_exclusion_count=sum(
                entry.entry_kind == "case_exclusion" for entry in self.entries
            ),
            entries_with_scenarios=sum(bool(entry.scenario_ids) for entry in self.entries),
        )
        if self.summary != expected:
            raise ValueError("failure discovery summary must be derived from entries")
        return self


def build_failure_discovery_index(
    repository: KnowledgeRepository,
    *,
    dataset_version: str,
    generated_at: datetime,
    gallery_index: dict[str, Any],
    learning_journeys: Any,
) -> FailureDiscoveryIndex:
    methods = {
        str(row["method_id"]): row
        for row in repository.fetch_all(
            "SELECT method_id, name_ja, name_en FROM methods ORDER BY method_id"
        )
    }
    failures = repository.structured_failure_modes()
    failure_ids_by_method: dict[str, set[str]] = {}
    for failure in failures:
        failure_id = str(failure["failure_mode_id"])
        for affected in failure["affected_entities"]:
            if affected["entity_type"] == "method":
                failure_ids_by_method.setdefault(str(affected["entity_id"]), set()).add(failure_id)

    journeys_by_case = {str(journey.case_id): journey for journey in learning_journeys.journeys}
    entries: list[FailureDiscoveryEntry] = []
    for failure in failures:
        affected = list(failure["affected_entities"])
        method_ids = sorted(
            str(item["entity_id"]) for item in affected if item["entity_type"] == "method"
        )
        implementation_ids = sorted(
            str(item["entity_id"]) for item in affected if item["entity_type"] == "implementation"
        )
        feature_ids = sorted(
            str(item["entity_id"]) for item in affected if item["entity_type"] == "feature"
        )
        failure_id = str(failure["failure_mode_id"])
        entries.append(
            FailureDiscoveryEntry(
                entry_id=f"structured:{failure_id}",
                entry_kind="structured_failure",
                disposition=(
                    "excluded" if str(failure["diagnose_disposition"]) == "exclude" else "warning"
                ),
                title_ja=str(failure["name_ja"]),
                title_en=str(failure["name_en"]),
                summary=str(failure["symptoms"][0]["description"]),
                scope=failure["failure_scope"],
                severity=failure["severity"],
                recoverability=failure["recoverability"],
                confidence=failure["confidence"],
                failure_mode_id=failure_id,
                case_id=None,
                method_ids=method_ids,
                implementation_ids=implementation_ids,
                feature_ids=feature_ids,
                scenario_ids=sorted(map(str, failure["scenario_ids"])),
                source_ids=sorted(map(str, failure["source_ids"])),
                symptoms=sorted(str(item["description"]) for item in failure["symptoms"]),
                diagnostics=sorted(str(item["check_text"]) for item in failure["diagnostics"]),
                mitigations=[
                    FailureMitigation(
                        action=str(item["action"]),
                        applicability=str(item["applicability"]),
                        tradeoff=str(item["tradeoff"]),
                    )
                    for item in sorted(
                        failure["mitigations"], key=lambda item: (item["priority"], item["action"])
                    )
                ],
                related_failure_mode_ids=[],
                case_context=None,
                canonical_route=f"/failures?entry={quote(f'structured:{failure_id}', safe='')}",
                last_verified=str(failure["last_verified"]),
            )
        )

    for case in sorted(gallery_index["cases"], key=lambda item: str(item["case_id"])):
        if case.get("status") != "published":
            continue
        case_id = str(case["case_id"])
        journey = journeys_by_case.get(case_id)
        scenario_ids = (
            sorted(str(item.scenario_id) for item in journey.scenarios)
            if journey is not None
            else []
        )
        context = CaseContext(
            question=str(case["question"]),
            decision_variables=str(case["decision_variables"]),
            objective=str(case["objective"]),
            constraints=str(case["constraints"]),
        )
        for exclusion in sorted(case["excluded_methods"], key=lambda item: str(item["method_id"])):
            method_id = str(exclusion["method_id"])
            method = methods.get(method_id)
            if method is None:
                raise ValueError(f"case exclusion references unknown method: {method_id}")
            entry_id = f"case:{case_id}:{method_id}"
            entries.append(
                FailureDiscoveryEntry(
                    entry_id=entry_id,
                    entry_kind="case_exclusion",
                    disposition="excluded",
                    title_ja=f"{case['title_ja']}では{method['name_ja']}を選ばない",
                    title_en=f"Why {method['name_en']} is excluded for {case['title_en']}",
                    summary=str(exclusion["reason"]),
                    scope="case_specific",
                    severity="not_applicable",
                    recoverability="not_applicable",
                    confidence="not_applicable",
                    failure_mode_id=None,
                    case_id=case_id,
                    method_ids=[method_id],
                    implementation_ids=[],
                    feature_ids=sorted(
                        {
                            str(item["feature_id"])
                            for item in case.get("feature_values", [])
                            if isinstance(item, dict) and item.get("feature_id")
                        }
                    ),
                    scenario_ids=scenario_ids,
                    source_ids=sorted(map(str, case["source_ids"])),
                    symptoms=[],
                    diagnostics=[],
                    mitigations=[],
                    related_failure_mode_ids=sorted(failure_ids_by_method.get(method_id, set())),
                    case_context=context,
                    canonical_route=f"/failures?entry={quote(entry_id, safe='')}",
                    last_verified=str(case["last_reviewed"]),
                )
            )

    entries.sort(key=lambda entry: entry.entry_id)
    return FailureDiscoveryIndex(
        dataset_version=dataset_version,
        generated_at=generated_at,
        summary=FailureDiscoverySummary(
            total_entries=len(entries),
            structured_failure_count=sum(
                entry.entry_kind == "structured_failure" for entry in entries
            ),
            case_exclusion_count=sum(entry.entry_kind == "case_exclusion" for entry in entries),
            entries_with_scenarios=sum(bool(entry.scenario_ids) for entry in entries),
        ),
        entries=entries,
    )
